clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3783,6 +3634,327 @@ static void handleSYCLIntelMaxWorkGroupSize(Sema &S, Decl *D,
+// by 'num_simd_work_items' attribute must evenly divide the index that
+// increments fastest in the 'reqd_work_group_size' attribute.
+//
+// The arguments to reqd_work_group_size are ordered based on which index
+// increments the fastest. In OpenCL, the first argument is the index that
+// increments the fastest, and in SYCL, the last argument is the index that
+// increments the fastest.
+static bool CheckWorkGroupSize(Sema &S, const Expr *NSWIValue,
+                               const Expr *RWGSXDim, const Expr *RWGSZDim) {
+  // If any of the operand is still value dependent, we can't test anything.
+  const auto *NSWIValueExpr = dyn_cast<ConstantExpr>(NSWIValue);
+  const auto *RWGSXDimExpr = dyn_cast<ConstantExpr>(RWGSXDim);
+  const auto *RWGSZDimExpr = dyn_cast<ConstantExpr>(RWGSZDim);
+
+  if (!NSWIValueExpr || !RWGSXDimExpr || !RWGSZDimExpr)
+    return false;
+
+  // Otherwise, check which argument increments the fastest
+  // in OpenCL vs SYCL mode.
+  unsigned WorkGroupSize =
+      S.getLangOpts().OpenCL
+          ? (RWGSXDimExpr->getResultAsAPSInt()).getZExtValue()
+          : (RWGSZDimExpr->getResultAsAPSInt()).getZExtValue();
+
+  // Check if the required work group size specified by 'num_simd_work_items'
+  // attribute must evenly divide the index that increments fastest in the
+  // 'reqd_work_group_size' attribute.
+  return WorkGroupSize % NSWIValueExpr->getResultAsAPSInt().getZExtValue() != 0;
+}
+
+void Sema::AddReqdWorkGroupSizeAttr(Decl *D, const AttributeCommonInfo &CI,
+                                    Expr *XDim, Expr *YDim, Expr *ZDim) {
+  // Returns nullptr if diagnosing, otherwise returns the original expression
+  // or the original expression converted to a constant expression.
+  auto CheckAndConvertArg = [&](Expr *E) -> Expr * {
+    // Check if the expression is not value dependent.
+    if (!E->isValueDependent()) {
+      llvm::APSInt ArgVal;
+      ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+      if (Res.isInvalid())
+        return nullptr;
+      E = Res.get();
+
+      // This attribute requires a strictly positive value.
+      if (ArgVal <= 0) {
+        Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+            << CI << /*positive*/ 0;
+        return nullptr;
+      }
+    }
+    return E;
+  };
+
+  // Check all three argument values, and if any are bad, bail out. This will
+  // convert the given expressions into constant expressions when possible.
+  XDim = CheckAndConvertArg(XDim);
+  YDim = CheckAndConvertArg(YDim);
+  ZDim = CheckAndConvertArg(ZDim);
+  if (!XDim || !YDim || !ZDim)
+    return;
+
+  // If the declaration has a ReqdWorkGroupSizeAttr, check to see if
+  // the attribute holds equal values to (1, 1, 1) in case the value of
+  // SYCLIntelMaxGlobalWorkDimAttr equals to 0.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+    if (InvalidWorkGroupSizeAttrs(DeclAttr->getValue(), XDim, YDim, ZDim)) {
```

Comment:
The comment above `InvalidWorkGroupSizeAttrs `says "Handles max_work_group_size attribute". This is incorrect right? Since `InvalidWorkGroupSizeAttrs` is called for not only `max_work_group_size`